Dbz 2782 sql server connector edit (#1971)
debezium/debezium: documentation/modules/ROOT/pages/connectors/sqlserver.adoc

toc::[]
endif::community[]

The {prodname} SQL Server connector captures row-level changes that occur in the schemas of a SQL Server database.

ifdef::product[]
For details about the {prodname} SQL Server connector and its use, see following topics: 

* xref:overview-of-debezium-sql-server-connector[]
* xref:how-debezium-sql-server-connectors-work[]
* xref:descriptions-of-debezium-sql-server-connector-data-change-events[]
* xref:how-debezium-SQL Server-connectors-map-data-types[]
* xref:setting-up-sql-server-for-use-with-the-debezium-sql-server-connector[]
* xref:deploying-debezium-sql-server-connectors[]
* xref:Monitoring {prodname} SQL Server connector performance[]
* xref:how-debezium-postgresql-connectors-handle-faults-and-problems[]

endif::product[]

The first time that the {prodname} SQL Server connector connects to a SQL Server database or cluster, it takes a consistent snapshot of the schemas in the database.

> After the initial snapshot is complete, the connector continuously captures row-level changes for `INSERT`, `UPDATE`, or `DELETE` operations that are committed to tracked SQL Server databases. 

Reviewer: @roldanbob Word `tracking` and `tracked` can be really confusing in SSQL Server context as SQL Server supports change data capture we rely upon and Change *Tracking* (https://docs.microsoft.com/en-us/sql/relational-databases/track-changes/about-change-tracking-sql-server?view=sql-server-ver15) which is similar but distinct feature

After you set up CDC on the source database, the connector can capture row-level `INSERT`, `UPDATE`, and `DELETE` operations  
that occur in the database.
The connector writes event records for each source table to a Kafka topic especially dedicated to that table.
One topic exists for each tracked table.
Client applications read the Kafka topics for the database tables that they follow, and can respond to the row-level events they they consume from those topics.

The first time that the connector connects to a SQL Server database or cluster, it takes a consistent snapshot of the schemas for all tables for which it is configured to capture changes,
and streams this state to Kafka.
After the snapshot is complete, the connector continuously captures subsequent row-level changes that occur.
By first establishing a consistent view of all of the data, the connector can continue reading without having lost any of the changes that were made while the snapshot was taking place.

The {prodname} SQL Server connector is tolerant of failures.
As the connector reads changes and produces events, it records the position in the database log (_LSN / Log Sequence Number_) for each event.

> If the connector stops for any reason (including communication failures, network problems, or crashes), after a restart the connector resumes reading the SQL Server _CDC_ tables from the last point that it read.

Reviewer 2: Worth mentioning that the offset is committed periodically, not after each change event. I.e. there may be some duplicates after a crash.

* xref:how-debezium-SQL Server-connectors-map-data-types[]
endif::product[]

// Type: concept
// ModuleID: how-debezium-sql-server-connectors-perform-database-snapshots
// Title: How {prodname} SQL Server connectors perform database snapshots
[[sqlserver-snapshots]]
=== Snapshots

SQL Server CDC is not designed to store a complete history of database changes.
For the {prodname} SQL Server connector to establish a baseline for the current state of the database, 
it uses a process called _snapshotting_.

You can configure how the connector creates snapshots.
By default, the connector's snapshot mode is set to `initial`. 
Based on this `initial` snapshot mode, the first time that the connector starts, it performs an initial _consistent snapshot_ of the database.
This initial snapshot captures the structure and data for any tables that match the criteria defined by the `include` and `exclude` properties that are configured for the connector (for example, `table.include.list`, `column.include.list`, `table.exclude.list`, and so forth).

When the connector creates a snapshot, it completes the following tasks:

1. Determines the tables to be captured.

> 2. Obtains a lock on each of the monitored tables to ensure that no structural changes can occur to any of the tables.

Reviewer: Here again the word `monitored` is used which was replaced in introduction. The terminologiy should be unified.

Reviewer 2: We should generally use "capture" (discussed with @TovaCohen some time ago). "monitoring" refers to (JMX) monitoring and "tracking", yeah, it's potentially confusing.

7. Records the successful completion of the snapshot in the connector offsets.

// Type: concept
// ModuleID: how-the-debezium-sql-server-connector-reads-change-data-tables
// Title: How {prodname} SQL Server connectors read change data tables
=== Reading the change data tables

When the connector first starts, it takes a structural snapshot of the structure of the captured tables
and persists this information to its internal database history topic.
The connector then identifies a change table for each source table, and completes the following steps.

1. For each change table, the connector read all of the changes that were created between the last stored maximum LSN and the current maximum LSN.
2. The connector sorts the changes that it reads in ascending order, based on the values of their commit LSN and change LSN.
This sorting ensures that the changes are replayed by {prodname} in the order in which they occurred in the database.
3. The connector passes the commit and change LSNs as offsets to Kafka Connect.

> 4. The connector store the maximum LSN and restarts the process from Step 1.

Reviewer 2: ```suggestion
4. The connector stores the maximum LSN and restarts the process from Step 1.
```

The first time that the connector connects to a SQL Server database or cluster, it takes a consistent snapshot of the schemas for all tables for which it is configured to capture changes,
and streams this state to Kafka.
After the snapshot is complete, the connector continuously captures subsequent row-level changes that occur.
By first establishing a consistent view of all of the data, the connector can continue reading without having lost any of the changes that were made while the snapshot was taking place.

The {prodname} SQL Server connector is tolerant of failures.
As the connector reads changes and produces events, it periodically records the position of events in the database log (_LSN / Log Sequence Number_).
If the connector stops for any reason (including communication failures, network problems, or crashes), after a restart the connector resumes reading the SQL Server _CDC_ tables from the last point that it read.

NOTE: Offsets are committed periodically. 
They are not committed at the time that a change event occurs. 
As a result, following an outage, duplicate events might be generated.

Fault tolerance also applies to snapshots. 

> That is, if the connector stops during a snapshot, the connector begins a news snapshot when it restarts.

Reviewer 2: ```suggestion
That is, if the connector stops during a snapshot, the connector begins a new snapshot when it restarts.
```

[[sqlserver-topic-names]]
=== Topic names

> The SQL Server connector writes events for all `INSERT`, `UPDATE`, and `DELETE` operations on a single table to a single Kafka topic. 

Reviewer 2: ```suggestion
The SQL Server connector writes events for all `INSERT`, `UPDATE`, and `DELETE` operations for a given table to a single Kafka topic. 
```

> The name of the Kafka topics always takes the form _serverName_._schemaName_._tableName_.

Reviewer 2: ```suggestion
The name of the Kafka topics by default takes the form _serverName_._schemaName_._tableName_.
```

(the logic table router can adjust this)

Author: Is this configured through [Custom topic auto creation]https://debezium.io/documentation/reference/1.3/configuration/topic-auto-create-config.html?

Reviewer: @gunnarmorling IMHO the naminig always takes the described form unless an SMT is used. Now the wording implies that it is something that can be change by configuration.

Reviewer 2: > Is this configured through Custom topic auto creation?

I meant the logical topic router, see here: https://debezium.io/documentation/reference/1.3/configuration/topic-routing.html. So it should be something like this:

> The name of the Kafka topics by default takes the form _serverName_._schemaName_._tableName_. If this default isn't suitable for your requirements, you can customize the topic naming by using the logical topic routing SMT (link)...

----

// Type: reference
// ModuleID: how-debezium-SQL Server-connectors-map-data-types
// Title: How {prodname} SQL Server connectors map data types
[[sqlserver-data-types]]
=== Data type mappings

The {prodname} SQL Server connector represents changes to table row data by producing events that are structured like the table in which the row exists.
Each event contains fields to represent the column values for the row. 
The way in which an event represents the column values for an operation depends on the SQL data type of the column. 
In the event, the connector maps the fields for each SQL Server data type to both a _literal type_ and a _semantic type_. 

The connector can map SQL Server data types to both _literal_ and _semantic_ types.

> Literal type:: Describes how the value is literally represented by using Kafka Connect schema types, namely `INT8`, `INT16`, `INT32`, `INT64`, `FLOAT32`, `FLOAT64`, `BOOLEAN`, `STRING`, `BYTES`, `ARRAY`, `MAP`, and `STRUCT`.

Reviewer: Probably needs an empty line above, otherwise don't render correctly.

Semantic type:: Describes how the Kafka Connect schema captures the _meaning_ of the field using the name of the Kafka Connect schema for the field.

ifdef::product[]
For more information about data type mappings, see the following sections:

> * xref:sqlserver-basic-values[]

Reviewer: Needs an empty line before?

[[setting-up-sqlserver]]
== Setting up SQL Server

For {prodname} to capture change events from SQL Server tables, a SQL Server administrator with the necessary privileges must first run a query to enable CDC on the database. 
The administrator must then enable CDC for each table to be tracked.   

> For details about setting up SQL Server for use with the {prodname} connector, see the following sections:

Reviewer: Should go into conditional block below

ifdef::product[]
* xref:enabling-cdc-on the-sql-server-database[]
* xref:enabling-cdc-on-a-sql-server-table[]
* xref:verifying-debezium-connector-access-to-the-cdc-table[]
* xref:debezium-sql-server-connector-on-azure[]
endif::product[]

> After CDC is applied, it captures all of the `INSERT`, `UPDATE`, and `DELETE` operations that are committed to tracked tables.

Reviewer 2: See comment elsewhere re "tracked". Worth a generic grep.

// Type: procedure
// ModuleID: enabling-cdc-on the-sql-server-database
=== Enabling CDC on the SQL Server database

Before you can enable CDC for a table, you must enable it for the SQL Server database.
A SQL Server administrator enables CDC by running a system stored procedure. 
System stored procedures can be run by using SQL Server Management Studio, or by using Transact-SQL.

.Prerequisites
* You are a member of the _sysadmin_ fixed server role for the SQL Server.
* You are a db_owner of the database. 
* The SQL Server Agent is running.  

NOTE: The SQL Server CDC feature tracks changes to user-created tables only. You cannot enable CDC on the SQL Server `master` database.

.Procedure

. From the *View* menu in SQL Server Management Studio, click *Template Explorer*.
. In the *Template Browser*, expand *SQL Server Templates*.
. Expand *Change Data Capture > Configuration* and then click *Enable Database for CDC*.  
. In the template, replace the database name in the `USE` statement with the name of the database that you want to enable for CDC.
. Run the stored procedure `sys.sp_cdc_enable_db` to enable the database for CDC.
+

> After the database is enabled for CDC, a schema with the name ‘cdc’ is created, along with a CDC user, metadata tables, and other system objects.

Reviewer: Wrong apostrophes?

The stored procedures can be run by using SQL Server Management Studio, or by using Transact-SQL.
SQL Server CDC must be enabled for every table that you want to track. 

.Prerequisites
* CDC is enabled on the SQL Server database.
* The SQL Server Agent is running.  
* You are a member of the `db_owner` fixed database role for the database. 

.Procedure
. From the *View* menu in SQL Server Management Studio, click *Template Explorer*.
. In the *Template Browser*, expand *SQL Server Templates*.
. Expand *Change Data Capture > Configuration*, and then click *Enable Table Specifying Filegroup Option*.  
. In the template, replace the table name in the `USE` statement with the name of the table that you want to track.
. Run the stored procedure `sys.sp_cdc_enable_table`.
+ 
The following example shows how to enable CDC for the table `MyTable`:
+
.Example: Enabling CDC for a SQL Server table
[source,sql]
----
USE MyDB
GO

EXEC sys.sp_cdc_enable_table
@source_schema = N'dbo',
@source_name   = N'MyTable', //<.>
@role_name     = N'MyRole',  //<.>
@filegroup_name = N'MyDB_CT',//<.>
@supports_net_changes = 0
GO
----

> <.> Specifies the name of the table that you want to track.

Reviewer 2: ```suggestion
<.> Specifies the name of the table that you want to capture.
```

<.> Specifies a role `MyRole` to which you can add users to whom you want to grant `SELECT` permission on the captured columns of the source table. 
Users in the `sysadmin` or `db_owner` role also have access to the specified change tables. Set the value of `@role_name` to `NULL`, to allow only members in the `sysadmin` or `db_owner` to have full access to captured information.
<.> Specifies the `filegroup` where SQL Server places the change table for the captured table. 
The named `filegroup` must already exist. 
It is best not to locate change tables in the same `filegroup` that you use for source tables.

// Type: procedure
// ModuleID: verifying-debezium-connector-access-to-the-cdc-table

> === Verifying that the user has access to the _CDC_ table

Reviewer: Emphasize around CDC can be removed

.Prerequisites
* You have `SELECT` permission on all of the captured columns of the capture instance. 
Members of the `db_owner` database role can view information for all of the defined capture instances.
* You have membership in any gating roles that are defined for the table information that the query includes. 

.Procedure

. From the *View* menu in SQL Server Management Studio, click *Object Explorer*.
. From the Object Explorer, expand *Databases*, and then expand your database object, for example, *MyDB*. 
. Expand *Programmability > Stored Procedures > System Stored Procedures*.
. Run the `sys.sp_cdc_help_change_data_capture` stored procedure to query the table.
+
Queries should not return empty results.
+
The following example runs the stored precedure `sys.sp_cdc_help_change_data_capture` on the database `MyDB`:
+
.Example: Querying a table for CDC configuration information
[source, sql]
----
USE MyDB;
GO
EXEC sys.sp_cdc_help_change_data_capture
GO
----
+
The query returns configuration information for each table in the database that is enabled for CDC and that contains change data that the caller is authorized to access.

> If the result is empty, verify that the user has privileges to access both the capture instance and the _CDC_ tables.

Reviewer: Same as above

If you are working with immutable containers, see link:https://hub.docker.com/r/debezium/[{prodname}'s container images] for Zookeeper, Kafka, SQL Server and Kafka Connect with the SQL Server connector already installed and ready to run. You can also xref:operations/openshift.adoc[run {prodname} on Kubernetes and OpenShift].
endif::community[]

> For details about deploying the {prodname} SQL Server connector, see the following topics:

Reviewer: Should be renderd for product only?

* xref:installing-debezium-mysql-connectors[]
* xref:debezium-sql-server-connector-configuration-example[]
* xref:adding-sql-server-connector-configuration-to-kafka-connect[]
* xref:descriptions-of-debezium-sql-server-connector-configuration-properties[]

ifdef::product[]
To deploy a {prodname} SQL Server connector, install the {prodname} SQL Server connector archive, configure the connector, and start the connector by adding its configuration to Kafka Connect. 

// Type: procedure
// Title: Installing {prodname} SQL Server connectors

> [id="installing-debezium-mysql-connectors"]

Reviewer: Not MySQL

Author: Oh! Good catch! I didn't notice that at all. Thank you!

Be sure to consult the {link-kafka-docs}.html[Kafka documentation] for all of the configuration properties for Kafka producers and consumers. (The SQL Server connector does use the {link-kafka-docs}.html#newconsumerconfigs[new consumer].)

// Type: assembly
// ModuleID: refreshing-capture-tables-after-a-schema-change

> // Title: URefreshing capture tables after a schema change 

Reviewer 2: ```suggestion
// Title: Refreshing capture tables after a schema change 
```

> // Title: URefreshing capture tables after a schema change 

Reviewer: Typo?

[[sqlserver-schema-evolution]]
=== Database schema evolution

When change data capture is enabled for a SQL Server table, as changes occur in the table, event records are persisted to a capture table on the server.
If you introduce a change in the structure of the source table change, for example, by adding a new column, that change is not dynamically reflected in the change table.
For as long as the capture table continues to use the outdated schema, the {prodname} connector is unable to emit data change events for the table correctly.
You must intervene to refresh the capture table to enable the connector to resume processing change events.   
 
Because of the way that CDC is implemented in SQL Server, you cannot use {prodname} to update capture tables.

> To refresh capture tables, one must be a SQL Server database operator with elevated privileges can.

Reviewer: Is this correct sentence?

Author: Thanks for catching that. I removed the extra word at the end of the sentence.

As a {prodname} user, you must coordinate tasks with the SQL Server database operator to complete the schema refresh and restore streaming to Kafka topics.    

You can use one of the following methods to update capture tables after a schema change:

* xref:cold-schema-updates[]. In cold schema updates, capture tables are updated after you stop {prodname}.

> * xref:hot-schema-updates[]. In hot schema updates, capture tables are updated while {prodname} is running.

Reviewer: Broken link

Author: Added missing `hot-schema-updates` anchor id.

// ModuleID: debezium-sql-server-connector-schema-history-metrics
// Title: {prodname} SQL Server connector schema history metrics
[[sqlserver-schema-history-metrics]]

> ==== Schema History Metrics

Reviewer 2: I think the title casing still is rather inconsistent. I don't think we used upper-casing most of the times. So perhaps refrain from doing it here, _or_ do it consistently everywhere?
